In [32]:
# !pip install tensorflow as tf

In [33]:
# @Secure_Voice_Channel
def generic_vns_function(input_dim, number_dense_layers, classes, units):

    # adjusted to effectively import the libraries
    from keras.models import Sequential
    from keras.layers import Dense
    model = Sequential()

    # adjusted to leverage the correct functions    
    for i in range(number_dense_layers):
        model.add(Dense(units=units, input_dim=input_dim, 
                               kernel_initializer='normal', 
                               activation='relu'))

    model.add(Dense(classes, kernel_initializer='normal',
                           activation='softmax'))
    model.compile(loss='categorical_crossentropy', optimizer='adam',
                  metrics=['accuracy'])
    return model

# adjusted to ingest varying number of epochs, batch sizes and units
def train_model(number_dense_layers, X_train, y_train, X_test, y_test,
                epochs, batch_size, units):
    # adjusted to leverage the generic_vns_function function
    model = generic_vns_function(X_train.shape[1], number_dense_layers,
                             y_train.shape[1], units)
    model.fit(X_train, y_train, validation_data=(X_test, y_test), 
              epochs=epochs, batch_size=batch_size, verbose=2)
    scores = model.evaluate(X_test, y_test, verbose=2)
    print("Baseline Error: %.2f%%" % (100-scores[1]*100))
    return model

#### Computer vision

In [26]:
# load data
from keras.datasets import mnist

(X_train, y_train), (X_test, y_test) = mnist.load_data()

In [27]:
# generic_vns_function(input_dim=num_pixels, number_dense_layers=1,
# @ classes=classes, units=1000)

In [28]:
# reshaping the dataset from 28x28 to 784
num_pixels = X_test.shape[1]*X_test.shape[2] # 28x28 pixels = 784
X_test = X_test.reshape(X_test.shape[0], num_pixels).astype('float32')
X_train = X_train.reshape(X_train.shape[0], num_pixels).astype('float32')

In [29]:
# normalize inputs
X_train = X_train / 255
X_test = X_test / 255

In [30]:
# one hot encode outputs
from tensorflow.keras.utils import to_categorical
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)
classes = y_test.shape[1] # Number of possible classes, = 10.

In [81]:
train_model(number_dense_layers=1, X_train= X_train, y_train=y_train,
 X_test=X_test, y_test= y_test, epochs=10, batch_size=200,
 units=1000)

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
 - 3s - loss: 0.2571 - accuracy: 0.9264 - val_loss: 0.1220 - val_accuracy: 0.9632
Epoch 2/10
 - 3s - loss: 0.1006 - accuracy: 0.9710 - val_loss: 0.0889 - val_accuracy: 0.9729
Epoch 3/10
 - 3s - loss: 0.0642 - accuracy: 0.9812 - val_loss: 0.0764 - val_accuracy: 0.9767
Epoch 4/10
 - 3s - loss: 0.0429 - accuracy: 0.9882 - val_loss: 0.0645 - val_accuracy: 0.9795
Epoch 5/10
 - 3s - loss: 0.0297 - accuracy: 0.9920 - val_loss: 0.0680 - val_accuracy: 0.9791
Epoch 6/10
 - 3s - loss: 0.0228 - accuracy: 0.9937 - val_loss: 0.0636 - val_accuracy: 0.9791
Epoch 7/10
 - 3s - loss: 0.0155 - accuracy: 0.9963 - val_loss: 0.0584 - val_accuracy: 0.9813
Epoch 8/10
 - 3s - loss: 0.0113 - accuracy: 0.9974 - val_loss: 0.0696 - val_accuracy: 0.9789
Epoch 9/10
 - 3s - loss: 0.0093 - accuracy: 0.9979 - val_loss: 0.0613 - val_accuracy: 0.9821
Epoch 10/10
 - 3s - loss: 0.0064 - accuracy: 0.9990 - val_loss: 0.0584 - val_accuracy: 0.9826
Baseline Error: 1.7

##### Vision Observations from default model
* Each epoch loss decreases, and accuracy increases on the training dataset
* There seems to be a bit of warning sign around epoch 8: accuracy on the training set went up, but accuracy on the validaiton set went down.

In [97]:
# adding an additional dense layer
train_model(number_dense_layers=2, X_train= X_train, y_train=y_train,
 X_test=X_test, y_test= y_test, epochs=10, batch_size=200,
 units=1000)

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
 - 6s - loss: 0.1973 - accuracy: 0.9403 - val_loss: 0.1067 - val_accuracy: 0.9649
Epoch 2/10
 - 6s - loss: 0.0693 - accuracy: 0.9786 - val_loss: 0.0746 - val_accuracy: 0.9769
Epoch 3/10
 - 6s - loss: 0.0436 - accuracy: 0.9865 - val_loss: 0.0637 - val_accuracy: 0.9789
Epoch 4/10
 - 6s - loss: 0.0290 - accuracy: 0.9905 - val_loss: 0.0696 - val_accuracy: 0.9800
Epoch 5/10
 - 6s - loss: 0.0206 - accuracy: 0.9938 - val_loss: 0.0805 - val_accuracy: 0.9767
Epoch 6/10
 - 6s - loss: 0.0189 - accuracy: 0.9938 - val_loss: 0.0768 - val_accuracy: 0.9782
Epoch 7/10
 - 6s - loss: 0.0151 - accuracy: 0.9951 - val_loss: 0.0946 - val_accuracy: 0.9762
Epoch 8/10
 - 6s - loss: 0.0143 - accuracy: 0.9953 - val_loss: 0.0883 - val_accuracy: 0.9783
Epoch 9/10
 - 6s - loss: 0.0155 - accuracy: 0.9950 - val_loss: 0.0689 - val_accuracy: 0.9827
Epoch 10/10
 - 6s - loss: 0.0113 - accuracy: 0.9965 - val_loss: 0.0858 - val_accuracy: 0.9794
Baseline Error: 2.0

Conclusion: adding an additional layer did not increase accuracy for the vision model

In [98]:
# increasing epochs
train_model(number_dense_layers=1, X_train= X_train, y_train=y_train,
 X_test=X_test, y_test= y_test, epochs=20, batch_size=200,
 units=1000)

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
 - 3s - loss: 0.2592 - accuracy: 0.9256 - val_loss: 0.1316 - val_accuracy: 0.9592
Epoch 2/20
 - 3s - loss: 0.1007 - accuracy: 0.9703 - val_loss: 0.0898 - val_accuracy: 0.9733
Epoch 3/20
 - 3s - loss: 0.0633 - accuracy: 0.9816 - val_loss: 0.0735 - val_accuracy: 0.9774
Epoch 4/20
 - 3s - loss: 0.0433 - accuracy: 0.9877 - val_loss: 0.0618 - val_accuracy: 0.9793
Epoch 5/20
 - 3s - loss: 0.0310 - accuracy: 0.9915 - val_loss: 0.0619 - val_accuracy: 0.9792
Epoch 6/20
 - 3s - loss: 0.0220 - accuracy: 0.9944 - val_loss: 0.0654 - val_accuracy: 0.9794
Epoch 7/20
 - 3s - loss: 0.0163 - accuracy: 0.9959 - val_loss: 0.0614 - val_accuracy: 0.9816
Epoch 8/20
 - 3s - loss: 0.0121 - accuracy: 0.9974 - val_loss: 0.0588 - val_accuracy: 0.9820
Epoch 9/20
 - 3s - loss: 0.0087 - accuracy: 0.9984 - val_loss: 0.0592 - val_accuracy: 0.9807
Epoch 10/20
 - 3s - loss: 0.0058 - accuracy: 0.9991 - val_loss: 0.0577 - val_accuracy: 0.9826
Epoch 11/20
 - 3s -

Conclusion: increasing the number of epochs does increase accuracy for the vision model

In [99]:
# decreasing batch size
train_model(number_dense_layers=1, X_train= X_train, y_train=y_train,
 X_test=X_test, y_test= y_test, epochs=20, batch_size=100,
 units=1000)

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
 - 5s - loss: 0.2231 - accuracy: 0.9347 - val_loss: 0.1150 - val_accuracy: 0.9665
Epoch 2/20
 - 5s - loss: 0.0854 - accuracy: 0.9747 - val_loss: 0.0791 - val_accuracy: 0.9753
Epoch 3/20
 - 5s - loss: 0.0536 - accuracy: 0.9835 - val_loss: 0.0673 - val_accuracy: 0.9795
Epoch 4/20
 - 5s - loss: 0.0344 - accuracy: 0.9898 - val_loss: 0.0714 - val_accuracy: 0.9776
Epoch 5/20
 - 5s - loss: 0.0247 - accuracy: 0.9930 - val_loss: 0.0616 - val_accuracy: 0.9800
Epoch 6/20
 - 5s - loss: 0.0180 - accuracy: 0.9948 - val_loss: 0.0618 - val_accuracy: 0.9812
Epoch 7/20
 - 5s - loss: 0.0137 - accuracy: 0.9959 - val_loss: 0.0679 - val_accuracy: 0.9807
Epoch 8/20
 - 5s - loss: 0.0104 - accuracy: 0.9970 - val_loss: 0.0662 - val_accuracy: 0.9819
Epoch 9/20
 - 5s - loss: 0.0105 - accuracy: 0.9969 - val_loss: 0.0617 - val_accuracy: 0.9825
Epoch 10/20
 - 5s - loss: 0.0075 - accuracy: 0.9981 - val_loss: 0.0654 - val_accuracy: 0.9818
Epoch 11/20
 - 5s -

Conclusion: decreasing batch size did not increase accuracy for the vision model


In [31]:
# increasing epochs
train_model(number_dense_layers=1, X_train= X_train, y_train=y_train,
 X_test=X_test, y_test= y_test, epochs=20, batch_size=200,
 units=2000)

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
 - 6s - loss: 0.2255 - accuracy: 0.9326 - val_loss: 0.1198 - val_accuracy: 0.9647
Epoch 2/20
 - 6s - loss: 0.0833 - accuracy: 0.9757 - val_loss: 0.0786 - val_accuracy: 0.9756
Epoch 3/20
 - 7s - loss: 0.0521 - accuracy: 0.9845 - val_loss: 0.0726 - val_accuracy: 0.9765
Epoch 4/20
 - 6s - loss: 0.0334 - accuracy: 0.9901 - val_loss: 0.0663 - val_accuracy: 0.9792
Epoch 5/20
 - 6s - loss: 0.0211 - accuracy: 0.9944 - val_loss: 0.0709 - val_accuracy: 0.9778
Epoch 6/20
 - 7s - loss: 0.0141 - accuracy: 0.9965 - val_loss: 0.0624 - val_accuracy: 0.9811
Epoch 7/20
 - 7s - loss: 0.0104 - accuracy: 0.9976 - val_loss: 0.0602 - val_accuracy: 0.9826
Epoch 8/20
 - 7s - loss: 0.0064 - accuracy: 0.9987 - val_loss: 0.0666 - val_accuracy: 0.9806
Epoch 9/20
 - 7s - loss: 0.0082 - accuracy: 0.9980 - val_loss: 0.0661 - val_accuracy: 0.9818
Epoch 10/20
 - 7s - loss: 0.0089 - accuracy: 0.9974 - val_loss: 0.0684 - val_accuracy: 0.9810
Epoch 11/20
 - 7s -

Conclusion:
* the best model turned to be a model with one dense layer and 20 epochs
* there doesn't seem to be an intuitive reason why neither increasing batch size or units can't improve the result. Setting hyperparameters so far seems more than an art than a science. 

### Speech recognition

In [7]:
pip install python_speech_features

Note: you may need to restart the kernel to use updated packages.


In [13]:
pip install pandas

     |████████████████████████████████| 10.4 MB 2.4 MB/s eta 0:00:01
  Using cached pytz-2020.1-py2.py3-none-any.whl (510 kB)
Note: you may need to restart the kernel to use updated packages.


In [ ]:
#import python_speech_features
# mfcc = python_speech_features.mfcc(file, rate)

In [11]:
# import spoken digit dataset (get_speech_dataset)
from db_utils import get_speech_dataset
(X_train, y_train), (X_test, y_test), (X_val, y_val) = get_speech_dataset()
# (X_train, y_train), (X_test, y_test) = get_speech_dataset()

In [12]:
# flatten the data
num_pixels = X_test.shape[1]*X_test.shape[2]
X_test = X_test.reshape(X_test.shape[0], num_pixels).astype('float32')
X_train = X_train.reshape(X_train.shape[0], num_pixels).astype('float32')

In [13]:
# standardize the inputs
import numpy as np
mean = np.mean(X_train)
std = np.std(X_train)
X_train = (X_train-std)/mean
X_test = (X_test-std)/mean

In [14]:
# one hot encode the outputs to avoid squeueing due to proximity of values
from tensorflow.keras.utils import to_categorical
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)
classes = y_test.shape[1] # Number of possible classes, = 10.

In [86]:
train_model(number_dense_layers=1, X_train= X_train, y_train=y_train,
 X_test=X_test, y_test= y_test, epochs=10, batch_size=200,
 units=1000)

Train on 18620 samples, validate on 2552 samples
Epoch 1/10
 - 3s - loss: 26.8836 - accuracy: 0.3362 - val_loss: 3.2326 - val_accuracy: 0.4189
Epoch 2/10
 - 3s - loss: 2.0473 - accuracy: 0.4948 - val_loss: 1.9372 - val_accuracy: 0.5106
Epoch 3/10
 - 3s - loss: 1.3198 - accuracy: 0.5996 - val_loss: 1.5285 - val_accuracy: 0.5803
Epoch 4/10
 - 3s - loss: 1.0028 - accuracy: 0.6780 - val_loss: 1.2918 - val_accuracy: 0.6364
Epoch 5/10
 - 3s - loss: 0.8033 - accuracy: 0.7401 - val_loss: 1.1430 - val_accuracy: 0.6642
Epoch 6/10
 - 3s - loss: 0.6661 - accuracy: 0.7844 - val_loss: 1.0823 - val_accuracy: 0.6947
Epoch 7/10
 - 3s - loss: 0.5638 - accuracy: 0.8145 - val_loss: 1.0386 - val_accuracy: 0.7092
Epoch 8/10
 - 3s - loss: 0.4798 - accuracy: 0.8439 - val_loss: 0.9570 - val_accuracy: 0.7386
Epoch 9/10
 - 3s - loss: 0.4087 - accuracy: 0.8653 - val_loss: 0.9334 - val_accuracy: 0.7496
Epoch 10/10
 - 3s - loss: 0.3387 - accuracy: 0.8905 - val_loss: 0.9283 - val_accuracy: 0.7645
Baseline Error: 23.

##### Speech - Conclusions on default model
* The model we developed generalizes between computer vision and speech recognition
* In contrast to computer vision, the baseline error remains 23.55%. 
* It is reasonable to assume running more epochs could help as both train and validation error are still decreasing
* It also seems reasonable to assume that other model specs could further improve the model.

In [104]:
# adding an additional dense layer
train_model(number_dense_layers=2, X_train= X_train, y_train=y_train,
 X_test=X_test, y_test= y_test, epochs=10, batch_size=200,
 units=1000)

Train on 18620 samples, validate on 2552 samples
Epoch 1/10
 - 4s - loss: 22.9293 - accuracy: 0.3347 - val_loss: 1.9929 - val_accuracy: 0.4412
Epoch 2/10
 - 4s - loss: 1.4581 - accuracy: 0.5432 - val_loss: 1.5184 - val_accuracy: 0.5517
Epoch 3/10
 - 4s - loss: 1.0155 - accuracy: 0.6739 - val_loss: 1.3201 - val_accuracy: 0.6148
Epoch 4/10
 - 4s - loss: 0.7683 - accuracy: 0.7487 - val_loss: 1.1635 - val_accuracy: 0.6646
Epoch 5/10
 - 4s - loss: 0.5859 - accuracy: 0.8044 - val_loss: 1.0080 - val_accuracy: 0.7104
Epoch 6/10
 - 4s - loss: 0.4675 - accuracy: 0.8420 - val_loss: 1.0629 - val_accuracy: 0.7132
Epoch 7/10
 - 4s - loss: 0.3709 - accuracy: 0.8748 - val_loss: 0.9579 - val_accuracy: 0.7382
Epoch 8/10
 - 4s - loss: 0.2747 - accuracy: 0.9086 - val_loss: 0.9272 - val_accuracy: 0.7582
Epoch 9/10
 - 4s - loss: 0.2252 - accuracy: 0.9231 - val_loss: 0.9172 - val_accuracy: 0.7680
Epoch 10/10
 - 4s - loss: 0.1838 - accuracy: 0.9376 - val_loss: 0.9212 - val_accuracy: 0.7751
Baseline Error: 22.

Conclusion: adding a layer improves the speech model

In [105]:
# increasing epochs while using one layer
train_model(number_dense_layers=1, X_train= X_train, y_train=y_train,
 X_test=X_test, y_test= y_test, epochs=20, batch_size=200,
 units=1000)

Train on 18620 samples, validate on 2552 samples
Epoch 1/20
 - 3s - loss: 28.4327 - accuracy: 0.3272 - val_loss: 3.2682 - val_accuracy: 0.4122
Epoch 2/20
 - 3s - loss: 2.0190 - accuracy: 0.4762 - val_loss: 1.8003 - val_accuracy: 0.4765
Epoch 3/20
 - 3s - loss: 1.3007 - accuracy: 0.5833 - val_loss: 1.4661 - val_accuracy: 0.5748
Epoch 4/20
 - 3s - loss: 1.0065 - accuracy: 0.6689 - val_loss: 1.3271 - val_accuracy: 0.6148
Epoch 5/20
 - 3s - loss: 0.8277 - accuracy: 0.7251 - val_loss: 1.2889 - val_accuracy: 0.6399
Epoch 6/20
 - 3s - loss: 0.7078 - accuracy: 0.7653 - val_loss: 1.0619 - val_accuracy: 0.6842
Epoch 7/20
 - 3s - loss: 0.5941 - accuracy: 0.8014 - val_loss: 1.0487 - val_accuracy: 0.6920
Epoch 8/20
 - 3s - loss: 0.4959 - accuracy: 0.8363 - val_loss: 0.9693 - val_accuracy: 0.7257
Epoch 9/20
 - 3s - loss: 0.4335 - accuracy: 0.8550 - val_loss: 0.9776 - val_accuracy: 0.7237
Epoch 10/20
 - 3s - loss: 0.3765 - accuracy: 0.8758 - val_loss: 0.9247 - val_accuracy: 0.7363
Epoch 11/20
 - 3s -

Conclusion: increasing epochs (without increasing the model) increases accuracy. The model also doesn't seem to be done learning, so we'll increase the number of epochs in the next model. We'll also combine more epochs with two layers.

In [7]:
# two layers; increased epochs
train_model(number_dense_layers=2, X_train= X_train, y_train=y_train,
 X_test=X_test, y_test= y_test, epochs=30, batch_size=200,
 units=1000)

Using TensorFlow backend.


Train on 18620 samples, validate on 2552 samples
Epoch 1/30
 - 4s - loss: 22.8555 - accuracy: 0.3086 - val_loss: 1.9293 - val_accuracy: 0.4306
Epoch 2/30
 - 4s - loss: 1.4847 - accuracy: 0.5312 - val_loss: 1.5164 - val_accuracy: 0.5525
Epoch 3/30
 - 4s - loss: 1.0594 - accuracy: 0.6601 - val_loss: 1.2104 - val_accuracy: 0.6309
Epoch 4/30
 - 4s - loss: 0.7857 - accuracy: 0.7429 - val_loss: 1.0193 - val_accuracy: 0.6799
Epoch 5/30
 - 4s - loss: 0.6033 - accuracy: 0.8061 - val_loss: 0.9331 - val_accuracy: 0.7261
Epoch 6/30
 - 4s - loss: 0.4828 - accuracy: 0.8413 - val_loss: 0.8836 - val_accuracy: 0.7390
Epoch 7/30
 - 4s - loss: 0.4173 - accuracy: 0.8633 - val_loss: 0.8749 - val_accuracy: 0.7520
Epoch 8/30
 - 4s - loss: 0.3311 - accuracy: 0.8908 - val_loss: 0.8320 - val_accuracy: 0.7696
Epoch 9/30
 - 4s - loss: 0.2485 - accuracy: 0.9173 - val_loss: 0.7892 - val_accuracy: 0.7958
Epoch 10/30
 - 4s - loss: 0.2153 - accuracy: 0.9266 - val_loss: 0.8490 - val_accuracy: 0.7739
Epoch 11/30
 - 4s -

Conclusion: model error dropped significantly from 20+ to 16+ percent.
Validation continues to go down consistently. We'll increase the epochs.

In [15]:
# adding more epochs as validation accuracy is still increasing
train_model(number_dense_layers=2, X_train= X_train, y_train=y_train,
 X_test=X_test, y_test= y_test, epochs=40, batch_size=200,
 units=1000)

Train on 18620 samples, validate on 2552 samples
Epoch 1/40
 - 4s - loss: 28.7320 - accuracy: 0.3277 - val_loss: 2.0146 - val_accuracy: 0.4224
Epoch 2/40
 - 4s - loss: 1.4722 - accuracy: 0.5302 - val_loss: 1.5860 - val_accuracy: 0.5278
Epoch 3/40
 - 4s - loss: 1.0978 - accuracy: 0.6434 - val_loss: 1.2976 - val_accuracy: 0.6101
Epoch 4/40
 - 4s - loss: 0.8248 - accuracy: 0.7290 - val_loss: 1.1451 - val_accuracy: 0.6681
Epoch 5/40
 - 4s - loss: 0.6575 - accuracy: 0.7840 - val_loss: 1.0502 - val_accuracy: 0.6967
Epoch 6/40
 - 4s - loss: 0.5269 - accuracy: 0.8267 - val_loss: 0.9702 - val_accuracy: 0.7183
Epoch 7/40
 - 4s - loss: 0.4284 - accuracy: 0.8612 - val_loss: 0.9807 - val_accuracy: 0.7226
Epoch 8/40
 - 4s - loss: 0.3497 - accuracy: 0.8838 - val_loss: 0.8753 - val_accuracy: 0.7712
Epoch 9/40
 - 4s - loss: 0.2888 - accuracy: 0.9050 - val_loss: 0.8916 - val_accuracy: 0.7641
Epoch 10/40
 - 4s - loss: 0.2471 - accuracy: 0.9156 - val_loss: 0.9434 - val_accuracy: 0.7641
Epoch 11/40
 - 4s -

Conclusion: we observe signs of overfitting (decreasing validation errors) as of epoch 35. We'll stop the next model there, and explore batch size next.

In [18]:
train_model(number_dense_layers=2, X_train= X_train, y_train=y_train,
 X_test=X_test, y_test= y_test, epochs=35, batch_size=100,
 units=1000)

Train on 18620 samples, validate on 2552 samples
Epoch 1/35
 - 7s - loss: 14.3605 - accuracy: 0.3959 - val_loss: 1.5848 - val_accuracy: 0.5223
Epoch 2/35
 - 6s - loss: 1.1489 - accuracy: 0.6321 - val_loss: 1.3096 - val_accuracy: 0.6105
Epoch 3/35
 - 8s - loss: 0.7934 - accuracy: 0.7478 - val_loss: 1.0109 - val_accuracy: 0.6928
Epoch 4/35
 - 7s - loss: 0.5902 - accuracy: 0.8072 - val_loss: 0.9551 - val_accuracy: 0.7198
Epoch 5/35
 - 7s - loss: 0.4674 - accuracy: 0.8468 - val_loss: 0.8564 - val_accuracy: 0.7582
Epoch 6/35
 - 7s - loss: 0.4090 - accuracy: 0.8645 - val_loss: 0.8032 - val_accuracy: 0.7782
Epoch 7/35
 - 7s - loss: 0.3383 - accuracy: 0.8861 - val_loss: 0.8636 - val_accuracy: 0.7766
Epoch 8/35
 - 6s - loss: 0.2861 - accuracy: 0.9038 - val_loss: 0.8317 - val_accuracy: 0.7947
Epoch 9/35
 - 6s - loss: 0.2448 - accuracy: 0.9194 - val_loss: 0.9595 - val_accuracy: 0.7829
Epoch 10/35
 - 6s - loss: 0.2617 - accuracy: 0.9142 - val_loss: 1.0185 - val_accuracy: 0.7731
Epoch 11/35
 - 6s -

Conclusion: lowering batch size seems to improve the model mildly.

In [20]:
train_model(number_dense_layers=2, X_train= X_train, y_train=y_train,
 X_test=X_test, y_test= y_test, epochs=35, batch_size=100,
 units=2000)

Train on 18620 samples, validate on 2552 samples
Epoch 1/35
 - 18s - loss: 51.2224 - accuracy: 0.3471 - val_loss: 1.8722 - val_accuracy: 0.4165
Epoch 2/35
 - 18s - loss: 1.4065 - accuracy: 0.5387 - val_loss: 1.5753 - val_accuracy: 0.5255
Epoch 3/35
 - 18s - loss: 1.0712 - accuracy: 0.6488 - val_loss: 1.2885 - val_accuracy: 0.6164
Epoch 4/35
 - 18s - loss: 0.8431 - accuracy: 0.7248 - val_loss: 1.2451 - val_accuracy: 0.6528
Epoch 5/35
 - 18s - loss: 0.7023 - accuracy: 0.7705 - val_loss: 1.1592 - val_accuracy: 0.6814
Epoch 6/35
 - 18s - loss: 0.5708 - accuracy: 0.8150 - val_loss: 1.1567 - val_accuracy: 0.7081
Epoch 7/35
 - 18s - loss: 0.5881 - accuracy: 0.8125 - val_loss: 1.1787 - val_accuracy: 0.7116
Epoch 8/35
 - 18s - loss: 0.5060 - accuracy: 0.8424 - val_loss: 1.1888 - val_accuracy: 0.7339
Epoch 9/35
 - 18s - loss: 0.5177 - accuracy: 0.8416 - val_loss: 1.1389 - val_accuracy: 0.7394
Epoch 10/35
 - 17s - loss: 0.4844 - accuracy: 0.8508 - val_loss: 1.0547 - val_accuracy: 0.7680
Epoch 11/

Conclusion: increasing units made the model worse. Maybe we should reduce?

In [23]:
train_model(number_dense_layers=2, X_train= X_train, y_train=y_train,
 X_test=X_test, y_test= y_test, epochs=35, batch_size=100,
 units=750)

Train on 18620 samples, validate on 2552 samples
Epoch 1/35
 - 5s - loss: 9.4356 - accuracy: 0.4048 - val_loss: 1.5659 - val_accuracy: 0.5157
Epoch 2/35
 - 5s - loss: 1.1549 - accuracy: 0.6263 - val_loss: 1.2942 - val_accuracy: 0.6074
Epoch 3/35
 - 5s - loss: 0.7875 - accuracy: 0.7426 - val_loss: 0.9384 - val_accuracy: 0.7288
Epoch 4/35
 - 5s - loss: 0.5932 - accuracy: 0.8032 - val_loss: 0.8631 - val_accuracy: 0.7484
Epoch 5/35
 - 5s - loss: 0.4716 - accuracy: 0.8456 - val_loss: 0.8662 - val_accuracy: 0.7594
Epoch 6/35
 - 5s - loss: 0.3770 - accuracy: 0.8750 - val_loss: 0.9100 - val_accuracy: 0.7567
Epoch 7/35
 - 5s - loss: 0.3466 - accuracy: 0.8868 - val_loss: 0.7817 - val_accuracy: 0.7978
Epoch 8/35
 - 5s - loss: 0.2738 - accuracy: 0.9085 - val_loss: 0.8506 - val_accuracy: 0.7892
Epoch 9/35
 - 5s - loss: 0.2450 - accuracy: 0.9175 - val_loss: 0.8997 - val_accuracy: 0.7900
Epoch 10/35
 - 5s - loss: 0.2533 - accuracy: 0.9166 - val_loss: 0.8475 - val_accuracy: 0.8123
Epoch 11/35
 - 5s - 

The best model remains the one with 2 layers and batch size 100
Could adding more layers help?

In [24]:
train_model(number_dense_layers=3, X_train= X_train, y_train=y_train,
 X_test=X_test, y_test= y_test, epochs=35, batch_size=100,
 units=2000)

Train on 18620 samples, validate on 2552 samples
Epoch 1/35
 - 24s - loss: 76.6539 - accuracy: 0.4060 - val_loss: 2.0274 - val_accuracy: 0.5196
Epoch 2/35
 - 23s - loss: 1.1814 - accuracy: 0.6670 - val_loss: 1.4257 - val_accuracy: 0.6399
Epoch 3/35
 - 26s - loss: 0.7189 - accuracy: 0.7777 - val_loss: 1.2475 - val_accuracy: 0.6963
Epoch 4/35
 - 26s - loss: 0.4779 - accuracy: 0.8485 - val_loss: 1.1640 - val_accuracy: 0.7183
Epoch 5/35
 - 25s - loss: 0.3239 - accuracy: 0.8917 - val_loss: 1.1322 - val_accuracy: 0.7426
Epoch 6/35
 - 25s - loss: 0.2731 - accuracy: 0.9070 - val_loss: 1.1351 - val_accuracy: 0.7586
Epoch 7/35
 - 23s - loss: 0.2297 - accuracy: 0.9241 - val_loss: 1.2070 - val_accuracy: 0.7539
Epoch 8/35
 - 23s - loss: 0.2151 - accuracy: 0.9277 - val_loss: 1.1571 - val_accuracy: 0.7798
Epoch 9/35
 - 26s - loss: 0.2649 - accuracy: 0.9170 - val_loss: 1.2437 - val_accuracy: 0.7637
Epoch 10/35
 - 25s - loss: 0.3291 - accuracy: 0.9045 - val_loss: 1.1487 - val_accuracy: 0.7763
Epoch 11/

#### Speech conclusion
* The best speech model was a 2-layer model with 35 epochs. We probably need to introduce a type of early stopping where the number of epochs is halted after observing one or two decreases in validation accuracy. 
* Although 1000 unit seems arbitray, neither decreasing, nor increase the number seemed to produce a more accurate model

#### Natural Language Processing: IMDb review sentiment analysis

In [8]:
from db_utils import get_imdb_dataset
(X_train, y_train), (X_test, y_test) = get_imdb_dataset()

In [88]:
train_model(number_dense_layers=1, X_train= X_train, y_train=y_train,
 X_test=X_test, y_test= y_test, epochs=10, batch_size=200,
 units=1000)

Train on 25000 samples, validate on 25000 samples
Epoch 1/10
 - 239s - loss: 0.3478 - accuracy: 0.8502 - val_loss: 0.2992 - val_accuracy: 0.8777
Epoch 2/10
 - 285s - loss: 0.0556 - accuracy: 0.9855 - val_loss: 0.3229 - val_accuracy: 0.8769
Epoch 3/10
 - 235s - loss: 0.0129 - accuracy: 0.9986 - val_loss: 0.3615 - val_accuracy: 0.8776
Epoch 4/10
 - 246s - loss: 0.0044 - accuracy: 0.9999 - val_loss: 0.3898 - val_accuracy: 0.8775
Epoch 5/10
 - 256s - loss: 0.0024 - accuracy: 1.0000 - val_loss: 0.4160 - val_accuracy: 0.8778
Epoch 6/10
 - 277s - loss: 0.0017 - accuracy: 1.0000 - val_loss: 0.4308 - val_accuracy: 0.8780
Epoch 7/10
 - 292s - loss: 8.7237e-04 - accuracy: 1.0000 - val_loss: 0.4467 - val_accuracy: 0.8773
Epoch 8/10
 - 292s - loss: 6.2179e-04 - accuracy: 1.0000 - val_loss: 0.4601 - val_accuracy: 0.8772
Epoch 9/10
 - 239s - loss: 4.6525e-04 - accuracy: 1.0000 - val_loss: 0.4723 - val_accuracy: 0.8767
Epoch 10/10
 - 246s - loss: 3.5998e-04 - accuracy: 1.0000 - val_loss: 0.4832 - val_

##### Conclusions
* Accuracy on the training set is 100% while accuracy on the validatsion set is only 87.7%
* It seems the model is overfitting. As such, longer training with an identical model doesn't seem reasonable. 

In [9]:
# adding an additional dense layer
train_model(number_dense_layers=2, X_train= X_train, y_train=y_train,
 X_test=X_test, y_test= y_test, epochs=10, batch_size=200,
 units=1000)

Train on 25000 samples, validate on 25000 samples
Epoch 1/10
 - 199s - loss: 0.4523 - accuracy: 0.8070 - val_loss: 0.2909 - val_accuracy: 0.8794
Epoch 2/10
 - 193s - loss: 0.0651 - accuracy: 0.9795 - val_loss: 0.3443 - val_accuracy: 0.8681
Epoch 3/10
 - 189s - loss: 0.0039 - accuracy: 0.9996 - val_loss: 0.4677 - val_accuracy: 0.8736
Epoch 4/10
 - 191s - loss: 3.6917e-04 - accuracy: 1.0000 - val_loss: 0.5322 - val_accuracy: 0.8720
Epoch 5/10
 - 187s - loss: 7.5469e-04 - accuracy: 1.0000 - val_loss: 0.5352 - val_accuracy: 0.8756
Epoch 6/10
 - 178s - loss: 8.8936e-05 - accuracy: 1.0000 - val_loss: 0.5627 - val_accuracy: 0.8756
Epoch 7/10
 - 182s - loss: 5.6146e-05 - accuracy: 1.0000 - val_loss: 0.5849 - val_accuracy: 0.8756
Epoch 8/10


KeyboardInterrupt: 

##### Conclusions on Natural Language Processing
* Accuracy on the training set is 100% while accuracy on the validatsion set is only 87.7%
* It seems the model is overfitting. As such, longer training with an identical model doesn't seem reasonable. 
* We therefore halted training.

##### Overall conclusions
* We ran an identical set of functions on a vision, speech and NLP model.
* The fact that a couple dozen lines of code can produce a model across a variety of use cases is pretty amazing
* Optimizing hyperparameters seems trial and error so far. Increasing the number of layers did improve speech but not vision. Increasing epochs seemed to have a positive effect on both vision and speech. But NLP seems to need a different model spec in order to improve performance.
* Training time can go up significantly, even with these small datasets. We tried the vision model on google collab, and training times decreased materially.